README.markdown

```diff
@@ -200,12 +200,18 @@ it:
 
        $ createdb goatcounter
        $ psql goatcounter -c '\i db/schema.pgsql'
-       $ goatcounter -db 'postgresql://dbname=goatcounter' migrate all
+       $ goatcounter -db 'postgresql://?dbname=goatcounter' migrate all
```

Comment:
I'm not sure if this is correct – PostgreSQL has two types of connection strings: `key=value` and URL-style `postgresql://servername/dbname?key=value`. For example these two are the same:

    user=bob password=secret host=1.2.3.4 port=5432 dbname=mydb sslmode=verify-full
    postgres://bob:secret@1.2.3.4:5432/mydb?sslmode=verify-full

GoatCounter accepts both, but I'm not sure if you really want to mix the two.

In which cases didn't things work for you?

Reply:
How would distinguish first the variant between sqlite and postgres as there is no `postgres://` prefix.

This is solution I deployed:

https://github.com/Mic92/dotfiles/blob/80b2853b0936e7584ee5281e9872c148e52e74a6/nixos/eve/modules/goatcounter.nix#L16

Reply:
The current method is to check if there's a space:

        // TODO: this check is less than ideal; maybe we should just try both?
        // "user=bob password=secret host=1.2.3.4 port=5432 dbname=mydb sslmode=verify-full"
        // "postgres://bob:secret@1.2.3.4:5432/mydb?sslmode=verify-full"
        if !strings.ContainsRune(conn, ' ') {
            conn = opt.Connect
        }
        dbx, exists, err = connectPostgreSQL(opt.Connect, opt.Create)

This is probably the problem – because with just `dbname=[..]` there's no space. In an older version it checked for a `/` character, and the example worked with that.

I'd really like both to work, as different people are familiar with different connection strings. I think the best fix it to resolve this.

Reply:
This sounds quite complex to me. Anyway I just reduced my documentation changes to unix sockets.

Reply:
Yeah; blame PostgreSQL for having two different types of connection strings I guess 😅

At any rate, this should be fixed now.